google/cloud/bigquery/dtypes.py

```diff
@@ -0,0 +1,247 @@
+        return len(self._ndarray)
+
+    @property
+    def shape(self):
+        return self._ndarray.shape
+
+    @property
+    def ndim(self) -> int:
+        return self._ndarray.ndim
+
+    @property
+    def size(self) -> int:
+        return self._ndarray.size
+
+    @property
+    def nbytes(self) -> int:
+        return self._ndarray.nbytes
+
+    def copy(self):
+        return self[:]
+
+    def repeat(self, n):
+        return self.__class__(self._ndarray.repeat(n), self._dtype)
+
+
+#
+###########################################################################
+
+
+class _BaseDtype(pandas.core.dtypes.base.ExtensionDtype):
+    na_value = NaT
+    kind = "o"
+    names = None
+
+    @classmethod
+    def construct_from_string(cls, name):
+        if name != cls.name:
+            raise TypeError()
+
+        return cls()
+
+
+class _BaseArray(OpsMixin, NDArrayBackedExtensionArray):
+    def __init__(self, values, dtype=None, copy: bool = False):
+        if not (
+            isinstance(values, numpy.ndarray) and values.dtype == numpy.dtype("<M8[us]")
+        ):
+            values = self.__ndarray(values)
+        elif copy:
+            values = values.copy()
+
+        super().__init__(values=values, dtype=values.dtype)
+
+    @classmethod
+    def __ndarray(cls, scalars):
+        return numpy.array(
+            [None if scalar is None else cls._datetime(scalar) for scalar in scalars],
+            "M8[us]",
+        )
+
+    @classmethod
+    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
+        if dtype is not None:
+            assert dtype.__class__ is cls.dtype.__class__
+        return cls(cls.__ndarray(scalars))
+
+    _from_sequence_of_strings = _from_sequence
+
+    def _cmp_method(self, other, op):
+        if type(other) != type(self):
+            return NotImplemented
+        return op(self._ndarray, other._ndarray)
+
+    def __setitem__(self, key, value):
+        if isinstance(key, slice):
+            _datetime = self._datetime
+            value = [_datetime(v) for v in value]
+        else:
+            value = self._datetime(value)
+        return super().__setitem__(key, value)
+
+
+@pandas.core.dtypes.dtypes.register_extension_dtype
+class TimeDtype(_BaseDtype):
+    """
+    Extension dtype for time data.
+    """
+
+    name = "time"
+    type = datetime.time
+
+    def construct_array_type(self):
+        return TimeArray
+
+
+class TimeArray(_BaseArray):
+    """
+    Pandas array type containing time data
+    """
+
+    # Data are stored as datetime64 values with a date of Jan 1, 1970
+
+    dtype = TimeDtype()
+    _epoch = datetime.datetime(1970, 1, 1)
+
+    @classmethod
+    def _datetime(cls, scalar):
+        if isinstance(scalar, datetime.time):
+            return datetime.datetime.combine(cls._epoch, scalar)
+        elif isinstance(scalar, str):
+            # iso string
+            h, m, s = map(float, scalar.split(":"))
+            s, us = divmod(s, 1)
```

Comment:
Mod 1? Isn't that just "s"?

I notice in INTERVAL that BigQuery doesn't always return the full 6 digits.

https://github.com/googleapis/python-bigquery/pull/840/files#diff-c8c849e0dbfe70fa3d5124f55d379f53fe4caff056addea0b981f679857bcfd5R82-R86

In fact, sometimes it doesn't have any partial seconds at all.

Maybe we should put some of the time parsing in a shared `_helpers` module to be used by both INTERVAL and TIME? https://github.com/googleapis/python-bigquery/pull/840/files#diff-f06352fa086344932b4922958dc31556b5cb0be634f640e2e009063955d1ca95R215

Reply:
If `s` is a float, `divmod(s, 1)` gives the whole and fractional parts.  We then multiply the fractional part by 1,000,000 to get the number of microseconds.  It doesn't matter how many digits are after the dot.

Reply:
> Maybe we should put some of the time parsing in a shared `_helpers` module to be used by both INTERVAL and TIME?

Sure.

Reply:
> > Maybe we should put some of the time parsing in a shared `_helpers` module to be used by both INTERVAL and TIME?
> 
> Sure.

Well, :)

- We're talking about 2 lines of code.
- Given that we might want to split out the dtype implementations, I think it's best to keep the dtype code decoupled.